adapters/adoppler/adoppler.go

```diff
@@ -0,0 +1,196 @@
+) (
+	[]*adapters.RequestData,
+	[]error,
+) {
+	if len(req.Imp) == 0 {
+		return nil, nil
+	}
+
+	var datas []*adapters.RequestData
+	for _, imp := range req.Imp {
+		ext, err := unmarshalExt(imp.Ext)
+		if err != nil {
+			return nil, []error{&errortypes.BadInput{err.Error()}}
+		}
+
+		var r openrtb.BidRequest = *req
+		r.ID = req.ID + "-" + ext.AdUnit
+		r.Imp = []openrtb.Imp{imp}
+
+		body, err := json.Marshal(r)
+		if err != nil {
+			return nil, []error{err}
+		}
+
+		uri := fmt.Sprintf("%s/processHeaderBid/%s",
+			ads.endpoint, ext.AdUnit)
+		headers := http.Header{}
+		headers.Add("Accept", "application/json")
+		headers.Add("Content-Type", "application/json;charset=utf-8")
+		headers.Add("X-OpenRTB-Version", "2.5")
+		data := &adapters.RequestData{
+			Method:  "POST",
+			Uri:     uri,
+			Body:    body,
+			Headers: headers,
+		}
+		datas = append(datas, data)
+	}
+
+	return datas, nil
+}
+
+func (ads *Adoppler) MakeBids(
+	intReq *openrtb.BidRequest,
+	extReq *adapters.RequestData,
+	resp *adapters.ResponseData,
+) (
+	*adapters.BidderResponse,
+	[]error,
+) {
+	if resp.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+	if resp.StatusCode == http.StatusBadRequest {
+		return nil, []error{&errortypes.BadInput{"bad request"}}
+	}
+	if resp.StatusCode != http.StatusOK {
+		err := &errortypes.BadServerResponse{
+			fmt.Sprintf("unexpected status: %d", resp.StatusCode),
+		}
+		return nil, []error{err}
+	}
+
+	var bidResp openrtb.BidResponse
+	err := json.Unmarshal(resp.Body, &bidResp)
+	if err != nil {
+		err := &errortypes.BadServerResponse{
+			fmt.Sprintf("invalid body: %s", err.Error()),
+		}
+		return nil, []error{err}
+	}
+
+	impTypes := make(map[string]openrtb_ext.BidType)
+	for _, imp := range intReq.Imp {
+		if imp.Banner != nil {
+			impTypes[imp.ID] = openrtb_ext.BidTypeBanner
+		} else if imp.Video != nil {
+			impTypes[imp.ID] = openrtb_ext.BidTypeVideo
+		} else if imp.Audio != nil {
+			impTypes[imp.ID] = openrtb_ext.BidTypeAudio
+		} else if imp.Native != nil {
+			impTypes[imp.ID] = openrtb_ext.BidTypeNative
+		}
+	}
+
+	var bids []*adapters.TypedBid
+	for _, seatBid := range bidResp.SeatBid {
+		for _, bid := range seatBid.Bid {
+			tp, ok := impTypes[bid.ImpID]
+			if !ok {
+				err := &errortypes.BadServerResponse{
+					fmt.Sprintf("uknown impid: %s", bid.ImpID),
+				}
+				return nil, []error{err}
+			}
+
+			var bidVideo *openrtb_ext.ExtBidPrebidVideo
+			if tp == openrtb_ext.BidTypeVideo {
+				adsExt, err := unmarshalAdsExt(bid.Ext)
+				if err != nil {
+					return nil, []error{&errortypes.BadServerResponse{err.Error()}}
+				}
+				if adsExt.Video == nil {
+					return nil, []error{&errortypes.BadServerResponse{
+						"$.seatbid.bid.ext.ads.video required",
+					}}
+				}
+				bidVideo = &openrtb_ext.ExtBidPrebidVideo{
+					Duration:        adsExt.Video.Duration,
+					PrimaryCategory: head(bid.Cat),
+				}
+			}
+			tb := &adapters.TypedBid{
+				Bid:      &bid,
+				BidType:  tp,
```

Comment:
> My understanding here is that PrebidServer wan't let request of unsupported type be passed to the adapter.
> Also, my understanding here that even if audio or native request comes with the response it can be processed successfully, adapter doesn't depend on request/response type itself. The goal was to prevent these types to be sent to exchange, because we do not support them now.
> I thought that `static/bidder-info/adoppler.yaml` configures PrebidServer in the way that it will not pass unsupported types to the adapter. Isn't it? If PrebidServer doesn't do that I can add a check in `MakeRequests()`.
I double checked and you are right, there's no need for the check. Thank you.